File: core/src/test/java/ch/dissem/bitmessage/ProofOfWorkServiceTest.java
/*
 * Copyright 2016 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage;

import ch.dissem.bitmessage.cryptography.bc.BouncyCryptography;
import ch.dissem.bitmessage.entity.BitmessageAddress;
import ch.dissem.bitmessage.entity.ObjectMessage;
import ch.dissem.bitmessage.entity.Plaintext;
import ch.dissem.bitmessage.entity.payload.Msg;
import ch.dissem.bitmessage.ports.*;
import ch.dissem.bitmessage.utils.Singleton;
import ch.dissem.bitmessage.utils.TestUtils;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Arrays;

import static ch.dissem.bitmessage.entity.Plaintext.Type.MSG;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Matchers.isNull;
import static org.mockito.Mockito.anyLong;
import static org.mockito.Mockito.*;

/**
 * @author Christian Basler
 */
public class ProofOfWorkServiceTest {
    private ProofOfWorkService proofOfWorkService;

    private Cryptography cryptography;
    @Mock
    private InternalContext ctx;
    @Mock
    private ProofOfWorkRepository proofOfWorkRepo;
    @Mock
    private Inventory inventory;
    @Mock
    private NetworkHandler networkHandler;
    @Mock
    private MessageRepository messageRepo;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        cryptography = spy(new BouncyCryptography());
        Singleton.initialize(cryptography);

        ctx = mock(InternalContext.class);
        when(ctx.getProofOfWorkRepository()).thenReturn(proofOfWorkRepo);
        when(ctx.getInventory()).thenReturn(inventory);
        when(ctx.getNetworkHandler()).thenReturn(networkHandler);
        when(ctx.getMessageRepository()).thenReturn(messageRepo);

        proofOfWorkService = new ProofOfWorkService();
        proofOfWorkService.setContext(ctx);
    }

    @Test
    public void ensureMissingProofOfWorkIsDone() {
        when(proofOfWorkRepo.getItems()).thenReturn(Arrays.asList(new byte[64]));
        when(proofOfWorkRepo.getItem(any(byte[].class))).thenReturn(new ProofOfWorkRepository.Item(null, 1001, 1002));
        doNothing().when(cryptography).doProofOfWork(any(ObjectMessage.class), anyLong(), anyLong(), any(ProofOfWorkEngine.Callback.class));

        proofOfWorkService.doMissingProofOfWork();

        verify(cryptography).doProofOfWork((ObjectMessage) isNull(), eq(1001L), eq(1002L),
                any(ProofOfWorkEngine.Callback.class));
    }

    @Test
    public void ensureCalculatedNonceIsStored() throws Exception {
        BitmessageAddress identity = TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8");
        BitmessageAddress address = TestUtils.loadContact();
        Plaintext plaintext = new Plaintext.Builder(MSG).from(identity).to(address).message("", "").build();
        ObjectMessage object = new ObjectMessage.Builder()
                .payload(new Msg(plaintext))
                .build();
        object.sign(identity.getPrivateKey());
        object.encrypt(address.getPubkey());
        byte[] initialHash = new byte[64];
        byte[] nonce = new byte[]{1, 2, 3, 4, 5, 6, 7, 8};

        when(proofOfWorkRepo.getItem(initialHash)).thenReturn(new ProofOfWorkRepository.Item(object, 1001, 1002));
        when(messageRepo.getMessage(initialHash)).thenReturn(plaintext);

        proofOfWorkService.onNonceCalculated(initialHash, nonce);

        verify(proofOfWorkRepo).removeObject(eq(initialHash));
        verify(inventory).storeObject(eq(object));
        verify(networkHandler).offer(eq(object.getInventoryVector()));
        assertThat(plaintext.getInventoryVector(), equalTo(object.getInventoryVector()));
    }
}

File: core/src/test/kotlin/ch/dissem/bitmessage/ProofOfWorkServiceTest.kt
/*
 * Copyright 2017 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage

import ch.dissem.bitmessage.cryptography.bc.BouncyCryptography
import ch.dissem.bitmessage.entity.ObjectMessage
import ch.dissem.bitmessage.entity.Plaintext
import ch.dissem.bitmessage.entity.Plaintext.Type.MSG
import ch.dissem.bitmessage.entity.payload.GenericPayload
import ch.dissem.bitmessage.entity.payload.Msg
import ch.dissem.bitmessage.ports.Cryptography
import ch.dissem.bitmessage.ports.ProofOfWorkEngine
import ch.dissem.bitmessage.ports.ProofOfWorkRepository
import ch.dissem.bitmessage.utils.Singleton
import ch.dissem.bitmessage.utils.TestUtils
import com.nhaarman.mockito_kotlin.*
import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import java.util.*
import kotlin.properties.Delegates

/**
 * @author Christian Basler
 */
class ProofOfWorkServiceTest {
    private var cryptography by Delegates.notNull<Cryptography>()
    private var ctx by Delegates.notNull<InternalContext>()

    private var obj by Delegates.notNull<ObjectMessage>()

    @Before
    fun setUp() {
        cryptography = spy(BouncyCryptography())
        Singleton.initialize(cryptography)

        ctx = TestUtils.mockedInternalContext(
            cryptography = cryptography
        )

        obj = ObjectMessage(
            expiresTime = 0,
            stream = 1,
            payload = GenericPayload(1, 1, kotlin.ByteArray(0)),
            type = 42,
            version = 42
        )
    }

    @Test
    fun `ensure missing proof of work is done`() {
        whenever(ctx.proofOfWorkRepository.getItems()).thenReturn(Arrays.asList<ByteArray>(ByteArray(64)))
        whenever(ctx.proofOfWorkRepository.getItem(any())).thenReturn(ProofOfWorkRepository.Item(obj, 1001, 1002))
        doNothing().whenever(cryptography).doProofOfWork(any(), any(), any(), any<ProofOfWorkEngine.Callback>())

        ctx.proofOfWorkService.doMissingProofOfWork(10)

        verify(cryptography, timeout(1000)).doProofOfWork(eq(obj), eq(1001L), eq(1002L), any<ProofOfWorkEngine.Callback>())
    }

    @Test
    fun `ensure calculated nonce is stored`() {
        val identity = TestUtils.loadIdentity("BM-2cSqjfJ8xK6UUn5Rw3RpdGQ9RsDkBhWnS8")
        val address = TestUtils.loadContact()
        val plaintext = Plaintext.Builder(MSG).from(identity).to(address).message("", "").build()
        val objectMessage = ObjectMessage(
            expiresTime = 0,
            stream = 1,
            payload = Msg(plaintext)
        )
        objectMessage.sign(identity.privateKey!!)
        objectMessage.encrypt(address.pubkey!!)
        val initialHash = ByteArray(64)
        val nonce = byteArrayOf(1, 2, 3, 4, 5, 6, 7, 8)

        whenever(ctx.proofOfWorkRepository.getItem(initialHash)).thenReturn(ProofOfWorkRepository.Item(objectMessage, 1001, 1002))
        whenever(ctx.messageRepository.getMessage(initialHash)).thenReturn(plaintext)

        ctx.proofOfWorkService.onNonceCalculated(initialHash, nonce)

        verify(ctx.proofOfWorkRepository).removeObject(eq(initialHash))
        verify(ctx.inventory).storeObject(eq(objectMessage))
        verify(ctx.networkHandler).offer(eq(objectMessage.inventoryVector))
        assertThat(plaintext.inventoryVector, equalTo(objectMessage.inventoryVector))
    }
}
